File: python/fecg_open/core/peak_detector/setDetectOptions.py
"""
setDetectOptions: Validate and complete an options dict for the SQI/peak-detection pipeline.

Mirrors setDetectOptions.m from the peak-detector source.
"""


def setDetectOptions(opt=None):
    """
    Apply defaults and validate an SQI/peak-detection options dict.

    Parameters
    ----------
    opt : dict or None
        Partial options dict. Missing keys receive their default values.
        Unknown keys are ignored with a warning. None is treated as {}.

    Returns
    -------
    opt : dict
        Completed, validated options dict with all fields present.

    Fields
    ------
    SIZE_WIND : int (positive even)
        SQI evaluation window length (seconds). Default: 10.
    LG_MED : int (>= 0)
        Half-width of sliding-minimum smoothing across SQI windows. Default: 3.
    REG_WIN : int (>= 1)
        SQI re-evaluation stride (seconds). Default: 1.
    THR : float
        Beat-matching tolerance (seconds). Default: 0.150.
    SQI_THR : float in [0, 1]
        Minimum SQI to accept a signal as good quality. Default: 0.8.
    JQRS_THRESH : float
        jqrs energy threshold (fraction of signal max). Default: 0.3.
    JQRS_REFRAC : float
        jqrs refractory period (seconds). Default: 0.25.
    JQRS_INTWIN_SZ : int
        jqrs integration window size (samples). Default: 7.
    JQRS_WINDOW : int
        jqrs processing sub-window size (seconds). Default: 15.
    ENABLE_OTHER_DETECTORS : int (0 or 1)
        Enable ABP/PPG detectors. Default: 0.
    SAVE_STUFF : int (0 or 1)
        Save intermediate results to disk. Default: 0.
    HALF_WIND : float (derived)
        SIZE_WIND / 2. Set automatically; not a valid input field.
    """
    defaults = {
        'SIZE_WIND':               10,
        'LG_MED':                  3,
        'REG_WIN':                 1,
        'THR':                     0.150,
        'SQI_THR':                 0.8,
        'JQRS_THRESH':             0.3,
        'JQRS_REFRAC':             0.25,
        'JQRS_INTWIN_SZ':          7,
        'JQRS_WINDOW':             15,
        'ENABLE_OTHER_DETECTORS':  0,
        'SAVE_STUFF':              0,
    }

    if opt is None:
        opt = {}

    for key, val in list(opt.items()):
        if key not in defaults:
            print(f'setDetectOptions: ignoring unrecognised option {key!r}')
            continue

        if key in ('SAVE_STUFF', 'ENABLE_OTHER_DETECTORS'):
            if val not in (0, 1):
                raise ValueError(f'setDetectOptions: {key} must be 0 or 1, got {val!r}')

        elif key == 'SQI_THR':
            if not isinstance(val, (int, float)) or val < 0 or val > 1:
                raise ValueError(f'setDetectOptions: {key} must be in [0, 1], got {val!r}')

        elif key == 'SIZE_WIND':
            if not isinstance(val, int) or val <= 0 or val % 2 != 0:
                raise ValueError(
                    f'setDetectOptions: {key} must be a positive even integer, got {val!r}')

        elif key in ('LG_MED', 'REG_WIN', 'JQRS_INTWIN_SZ', 'JQRS_WINDOW'):
            if not isinstance(val, int):
                raise ValueError(f'setDetectOptions: {key} must be an integer, got {val!r}')

        elif key in ('THR', 'JQRS_THRESH', 'JQRS_REFRAC'):
            if not isinstance(val, (int, float)):
                raise ValueError(f'setDetectOptions: {key} must be numeric, got {val!r}')

        defaults[key] = val

    defaults['HALF_WIND'] = defaults['SIZE_WIND'] / 2
    return defaults

File: matlab/fecg_open/core/peak_detector/setDetectOptions.m
% SOURCE: https://raw.githubusercontent.com/alistairewj/peak-detector/master/sources/setDetectOptions.m
function [opt] = setDetectOptions(varargin)
% setDetectOptions  Build and validate the options struct for beat detection and SQI scoring.
%
% PURPOSE
%   Returns a struct of parameters used by the jQRS/gQRS beat detector and
%   the beat-quality index (bSQI) scorer.  When called with no arguments,
%   returns factory defaults.  When called with a struct, merges the
%   supplied fields into the defaults after field-by-field validation.
%
% INPUTS
%   varargin{1} - (optional struct) partial options struct.  Recognised
%                 fields and their defaults:
%                   SIZE_WIND             10   (even integer) SQI window, seconds
%                   LG_MED                 3   (integer)      SQI median-filter half-width
%                   REG_WIN                1   (integer)      jQRS regularisation window
%                   THR                0.150   (numeric)      jQRS beat-detection threshold
%                   SQI_THR              0.8   (0-1)          minimum bSQI to accept a segment
%                   JQRS_THRESH          0.3   (numeric)      jQRS peak-detection threshold
%                   JQRS_REFRAC         0.25   (numeric)      jQRS refractory period (s)
%                   JQRS_INTWIN_SZ         7   (integer)      jQRS integration window size
%                   JQRS_WINDOW           15   (integer)      jQRS search window (s)
%                   ENABLE_OTHER_DETECTORS 0   (0|1)          enable auxiliary detectors
%                   SAVE_STUFF             0   (0|1)          save intermediate detector output
%                 Unrecognised fields are silently ignored.
%
% OUTPUTS
%   opt - (struct) validated options struct.  Includes all fields above plus
%         the derived field HALF_WIND = SIZE_WIND/2.
%
%   Cite: Johnson et al. (2015), Physiological Measurement 36:1665-1677.

    opt_default.SIZE_WIND = 10;
    opt_default.LG_MED = 3;
    opt_default.REG_WIN = 1;
    opt_default.THR = 0.150;
    opt_default.SQI_THR = 0.8;
    opt_default.JQRS_THRESH = 0.3;
    opt_default.JQRS_REFRAC = 0.25;
    opt_default.JQRS_INTWIN_SZ = 7;
    opt_default.JQRS_WINDOW = 15;
    opt_default.ENABLE_OTHER_DETECTORS = 0;
    opt_default.SAVE_STUFF = 0;

    if nargin == 0
        opt = opt_default;
        return;
    elseif nargin == 1
        opt = varargin{1};
    else
        error('Incorrect number of inputs.');
    end

    if nargin > 0 && isstruct(opt)
        fn = fieldnames(opt);
        fn_default = fieldnames(opt_default);
        for f = 1:numel(fn)
            if ismember(fn{f}, fn_default)
                val = opt.(fn{f});

                switch fn{f}
                    case {'SAVE_STUFF', 'ENABLE_OTHER_DETECTORS'}
                        if ~isnumeric(val) || (val ~= 0 && val ~= 1)
                            error('setDetectOptions:badValue', '%s: field can only take values 0 or 1.', fn{f});
                        end
                    case {'THR', 'JQRS_THRESH', 'JQRS_REFRAC'}
                        if ~isnumeric(val)
                            error('setDetectOptions:badValue', '%s: field should be an integer.', fn{f});
                        end
                    case {'SQI_THR'}
                        if ~isnumeric(val) || val < 0 || val > 1
                            error('setDetectOptions:badValue', '%s: field should be an integer.', fn{f});
                        end
                    case {'LG_MED', 'REG_WIN', 'JQRS_INTWIN_SZ', 'JQRS_WINDOW'}
                        if ~isnumeric(val) || round(val) ~= val
                            error('setDetectOptions:badValue', '%s: field should be an integer.', fn{f});
                        end
                    case {'SIZE_WIND'}
                        if ~isnumeric(val) || round(val) ~= val || mod(val, 2) ~= 0
                            error('setDetectOptions:badValue', '%s: field should be an integer divisible by two.', fn{f});
                        end
                end

                opt_default.(fn{f}) = val;
            else
                fprintf('Ignoring unrecognized option %s\n', fn{f});
            end
        end
    end

    opt_default.HALF_WIND = opt_default.SIZE_WIND/2;
    opt = opt_default;

end
